File: evidence_record_print.lua
--[[
MIT License

Copyright (c) 2024 Florian Fischer

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
]]--

-- Visualisieren der Hashbäume nach Evidence Record Implementation nach RFC4998
local PrintEr = {}

-- Hilfsfunktion zum Erzeugen von Leerzeichen
local function spaces(n)
    return string.rep(" ", n)
end

-- Visualisiert einen einzelnen Knoten
local function visualizeNode(hash)
    local width = #hash
    local top = "+" .. string.rep("-", width + 2) .. "+"
    local middle = "| " .. hash .. " |"
    return {top, middle, top}
end

-- Hilfsfunktion zum Verbinden von Linien
local function connectNodes(leftLines, rightLines, centerSpacing)
    local result = {}
    -- Erste Zeile: Verbindungslinien
    local firstLine = ""
    local leftWidth = #(leftLines and leftLines[1] or "")
    local rightWidth = #(rightLines and rightLines[1] or "")
    
    if leftLines and rightLines then
        firstLine = spaces(leftWidth/2) .. "/" .. spaces(centerSpacing-2) .. "\\" .. spaces(rightWidth/2)
    elseif leftLines then
        firstLine = spaces(leftWidth/2) .. "|"
    elseif rightLines then
        firstLine = spaces(centerSpacing) .. spaces(rightWidth/2) .. "|"
    end
    
    table.insert(result, firstLine)
    return result
end

-- Visualisiert einen Hashbaum
function PrintEr.visualizeTree(node, prefix)
    if not node then return {} end
    prefix = prefix or ""
    
    local lines = {}
    local nodeLines = visualizeNode(node.hash)
    
    -- Rekursiv die Kindknoten visualisieren
    local leftLines = PrintEr.visualizeTree(node.left, prefix .. "  ")
    local rightLines = PrintEr.visualizeTree(node.right, prefix .. "  ")
    
    -- Breiten berechnen
    local leftWidth = leftLines[1] and #leftLines[1] or 0
    local rightWidth = rightLines[1] and #rightLines[1] or 0
    local nodeWidth = #nodeLines[1]
    local spacing = 4
    
    -- Knoten zentrieren
    local nodePadding = math.max(0, (leftWidth + rightWidth + spacing - nodeWidth) / 2)
    for _, line in ipairs(nodeLines) do
        table.insert(lines, spaces(math.floor(nodePadding)) .. line)
    end
    
    -- Verbindungslinien hinzufügen
    if node.left or node.right then
        local connections = connectNodes(leftLines[1], rightLines[1], spacing)
        for _, line in ipairs(connections) do
            table.insert(lines, line)
        end
    end
    
    -- Kindknoten hinzufügen
    local maxChildLines = math.max(#leftLines, #rightLines)
    for i = 1, maxChildLines do
        local leftLine = leftLines[i] or spaces(leftWidth)
        local rightLine = rightLines[i] or spaces(rightWidth)
        table.insert(lines, leftLine .. spaces(spacing) .. rightLine)
    end
    
    return lines
end

-- Helper function to print the lines to console
function PrintEr.displayTree(tree)
	local lines = PrintEr.visualizeTree(tree)
	for _, line in ipairs(lines) do
		print(line)
	end
end

-- Helper function to create a header box
local function createHeaderBox(text, width)
    local line = string.rep("-", width-2)
    return {
        "+" .. line .. "+",
        "| " .. text .. string.rep(" ", width-4-#text) .. " |",
        "+" .. line .. "+"
    }
end

-- Helper function to merge lines side by side
local function mergeLinesHorizontally(left, right, spacing)
    spacing = spacing or 4
    local result = {}
    local maxLines = math.max(#left, #right)
    
    for i = 1, maxLines do
        local leftLine = left[i] or string.rep(" ", #(left[1] or ""))
        local rightLine = right[i] or string.rep(" ", #(right[1] or ""))
        table.insert(result, leftLine .. string.rep(" ", spacing) .. rightLine)
    end
    
    return result
end

-- Print an Evidence Record structure
function PrintEr.printEvidenceRecord(evidenceRecord)
    if not evidenceRecord then
        return {"No evidence record provided"}
    end
    
    local lines = {}
    
    -- Print header information
    local headerInfo = createHeaderBox(string.format(
        "Evidence Record v%d - %s", 
        evidenceRecord.version,
        evidenceRecord.digestAlgorithm or "Unknown Algorithm"
    ), 60)
    
    for _, line in ipairs(headerInfo) do
        table.insert(lines, line)
    end
    table.insert(lines, "")
    
    -- Print Archive Time Stamp Sequence
    if evidenceRecord.archiveTimeStampSequence then
        table.insert(lines, "Archive Time Stamp Sequence:")
        table.insert(lines, string.rep("-", 30))
        
        for i, ats in ipairs(evidenceRecord.archiveTimeStampSequence) do
            -- Create chain header
            table.insert(lines, string.format("Chain %d:", i))
            table.insert(lines, "")
            
            -- Print timestamp information
            if ats.timestamp then
                local timestampBox = createHeaderBox(string.format(
                    "Timestamp: %s [%s]",
                    os.date("%Y-%m-%d %H:%M:%S", ats.timestamp.time),
                    ats.timestamp.algorithm or "Unknown"
                ), 50)
                
                for _, line in ipairs(timestampBox) do
                    table.insert(lines, "  " .. line)  -- indent timestamp
                end
            end

            -- Visualize reduced tree if present
            if ats.reduced then
                local treeLines = PrintEr.visualizeTree(ats.reduced)
                for _, line in ipairs(treeLines) do
                    table.insert(lines, "  " .. line)  -- indent tree
                end
            end
            
            table.insert(lines, "")
            if i < #evidenceRecord.archiveTimeStampSequence then
                table.insert(lines, string.rep("-", 30))
            end
        end
    end
    
    -- Return all lines joined by newlines
    return lines
end

-- Helper function to print the lines to console
function PrintEr.displayEvidenceRecord(evidenceRecord)
    local lines = PrintEr.printEvidenceRecord(evidenceRecord)
    for _, line in ipairs(lines) do
        print(line)
    end
end

return PrintEr

File: evidence_record_print.py
"""
MIT License

Copyright (c) 2024 Florian Fischer

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

from datetime import datetime
import time

class PrintEr:
    def spaces(self, n):
        """Helper function to generate spaces"""
        return " " * n

    def visualize_node(self, hash_value):
        """Visualizes a single node"""
        width = len(hash_value)
        top = f"+{'-' * (width + 2)}+"
        middle = f"| {hash_value} |"
        return [top, middle, top]

    def connect_nodes(self, left_lines, right_lines, center_spacing):
        """Helper function to connect lines"""
        result = []
        # First line: Connection lines
        first_line = ""
        left_width = len(left_lines[0]) if left_lines else 0
        right_width = len(right_lines[0]) if right_lines else 0

        if left_lines and right_lines:
            first_line = f"{self.spaces(left_width // 2)}/{self.spaces(center_spacing - 2)}\\{self.spaces(right_width // 2)}"
        elif left_lines:
            first_line = f"{self.spaces(left_width // 2)}|"
        elif right_lines:
            first_line = f"{self.spaces(center_spacing)}{self.spaces(right_width // 2)}|"

        result.append(first_line)
        return result

    def visualize_tree(self, node, prefix=""):
        """Visualizes a hash tree"""
        if not node:
            return []

        lines = []
        node_lines = self.visualize_node(node['hash'])

        # Recursively visualize child nodes
        left_lines = self.visualize_tree(node.get('left'), prefix + "  ")
        right_lines = self.visualize_tree(node.get('right'), prefix + "  ")

        # Calculate widths
        left_width = len(left_lines[0]) if left_lines else 0
        right_width = len(right_lines[0]) if right_lines else 0
        node_width = len(node_lines[0])
        spacing = 4

        # Center the node
        node_padding = max(0, (left_width + right_width + spacing - node_width) // 2)
        for line in node_lines:
            lines.append(self.spaces(node_padding) + line)

        # Add connection lines
        if node.get('left') or node.get('right'):
            connections = self.connect_nodes(
                left_lines[0] if left_lines else None,
                right_lines[0] if right_lines else None,
                spacing
            )
            lines.extend(connections)

        # Add child nodes
        max_child_lines = max(len(left_lines), len(right_lines))
        for i in range(max_child_lines):
            left_line = left_lines[i] if left_lines and i < len(left_lines) else self.spaces(left_width)
            right_line = right_lines[i] if right_lines and i < len(right_lines) else self.spaces(right_width)
            lines.append(left_line + self.spaces(spacing) + right_line)

        return lines

    def create_header_box(self, text, width):
        """Helper function to create a header box"""
        line = "-" * (width - 2)
        return [
            f"+{line}+",
            f"| {text}{self.spaces(width - 4 - len(text))} |",
            f"+{line}+"
        ]

    def merge_lines_horizontally(self, left, right, spacing=4):
        """Helper function to merge lines side by side"""
        result = []
        max_lines = max(len(left), len(right))
        left_width = len(left[0]) if left else 0
        
        for i in range(max_lines):
            left_line = left[i] if i < len(left) else self.spaces(left_width)
            right_line = right[i] if i < len(right) else self.spaces(len(right[0]) if right else 0)
            result.append(left_line + self.spaces(spacing) + right_line)
            
        return result

    def print_evidence_record(self, evidence_record):
        """Print an Evidence Record structure"""
        if not evidence_record:
            return ["No evidence record provided"]
        
        lines = []
        
        # Print header information
        header_info = self.create_header_box(
            f"Evidence Record v{evidence_record.get('version')} - {evidence_record.get('digestAlgorithm', 'Unknown Algorithm')}",
            60
        )
        
        lines.extend(header_info)
        lines.append("")
        
        # Print Archive Time Stamp Sequence
        ats_sequence = evidence_record.get('archiveTimeStampSequence', [])
        if ats_sequence:
            lines.append("Archive Time Stamp Sequence:")
            lines.append("-" * 30)
            
            for i, ats in enumerate(ats_sequence, 1):
                # Create chain header
                lines.append(f"Chain {i}:")
                lines.append("")
                
                # Print timestamp information
                if timestamp := ats.get('timestamp'):
                    timestamp_str = datetime.fromtimestamp(timestamp.get('time')).strftime("%Y-%m-%d %H:%M:%S")
                    timestamp_box = self.create_header_box(
                        f"Timestamp: {timestamp_str} [{timestamp.get('algorithm', 'Unknown')}]",
                        50
                    )
                    
                    for line in timestamp_box:
                        lines.append("  " + line)  # indent timestamp
                
                # Visualize reduced tree if present
                if reduced := ats.get('reduced'):
                    tree_lines = self.visualize_tree(reduced)
                    for line in tree_lines:
                        lines.append("  " + line)  # indent tree
                
                lines.append("")
                if i < len(ats_sequence):
                    lines.append("-" * 30)
        
        return lines

    def display_evidence_record(self, evidence_record):
        """Helper function to print the lines to console"""
        for line in self.print_evidence_record(evidence_record):
            print(line)
